### PR TITLE
[GVN][NewGVN][Local] Handle attributes for function calls after CSE

### LINKED ISSUE #113997: [GVNPass] Range attribute should be handled after CSE
Reproducer: https://alive2.llvm.org/ce/z/HAnoKn
```
define i1 @src(i32 noundef %x) {
  %cmp1 = icmp sgt i32 %x, 0
  %ctpop1 = tail call range(i32 1, 32) i32 @llvm.ctpop.i32(i32 %x)
  %cmp2 = icmp samesign ult i32 %ctpop1, 2
  %cond = select i1 %cmp1, i1 %cmp2, i1 false
  br i1 %cond, label %if.then, label %if.else

if.else:
  %ctpop2 = tail call range(i32 0, 33) i32 @llvm.ctpop.i32(i32 %x)
  %res = icmp eq i32 %ctpop2, 1
  ret i1 %res

if.then:
  ret i1 false
}

define i1 @tgt(i32 noundef %x) {
  %cmp1 = icmp sgt i32 %x, 0
  %ctpop1 = tail call range(i32 1, 32) i32 @llvm.ctpop.i32(i32 %x)
  %cmp2 = icmp samesign ult i32 %ctpop1, 2
  %cond = select i1 %cmp1, i1 %cmp2, i1 false
  br i1 %cond, label %if.then, label %if.else

if.else:
  %res = icmp eq i32 %ctpop1, 1
  ret i1 %res

if.then:
  ret i1 false
}
```
```

----------------------------------------
define i1 @src(i32 noundef %x) {
#0:
  %cmp1 = icmp sgt i32 noundef %x, 0
  %ctpop1 = ctpop i32 noundef %x
  %#range_0_%ctpop1 = !range i32 %ctpop1, i32 1, i32 32
  %cmp2 = icmp samesign ult i32 %#range_0_%ctpop1, 2
  %cond = select i1 %cmp1, i1 %cmp2, i1 0
  br i1 %cond, label %if.then, label %if.else

if.else:
  %ctpop2 = ctpop i32 noundef %x
  %#range_1_%ctpop2 = !range i32 %ctpop2, i32 0, i32 33
  %res = icmp eq i32 %#range_1_%ctpop2, 1
  ret i1 %res

if.then:
  ret i1 0
}
=>
define i1 @tgt(i32 noundef %x) {
#0:
  %cmp1 = icmp sgt i32 noundef %x, 0
  %ctpop1 = ctpop i32 noundef %x
  %#range_0_%ctpop1 = !range i32 %ctpop1, i32 1, i32 32
  %cmp2 = icmp samesign ult i32 %#range_0_%ctpop1, 2
  %cond = select i1 %cmp1, i1 %cmp2, i1 0
  br i1 %cond, label %if.then, label %if.else

if.else:
  %res = icmp eq i32 %#range_0_%ctpop1, 1
  ret i1 %res

if.then:
  ret i1 0
}
Transformation doesn't verify!

ERROR: Target is more poisonous than source

Example:
i32 noundef %x = #x00000000 (0)

Source:
i1 %cmp1 = #x0 (0)
i32 %ctpop1 = #x00000000 (0)
i32 %#range_0_%ctpop1 = poison
i1 %cmp2 = poison
i1 %cond = #x0 (0)
  >> Jump to %if.else
i32 %ctpop2 = #x00000000 (0)
i32 %#range_1_%ctpop2 = #x00000000 (0)
i1 %res = #x0 (0)

Target:
i1 %cmp1 = #x0 (0)
i32 %ctpop1 = #x00000000 (0)
i32 %#range_0_%ctpop1 = poison
i1 %cmp2 = poison
i1 %cond = #x0 (0)
  >> Jump to %if.else
i1 %res = poison
Source value: #x0 (0)
Target value: poison
```

Reported by @pranavk. See https://github.com/llvm/llvm-project/pull/100899#issuecomment-2442780517.

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch intersects attributes of two calls to avoid introducing UB. It also skips incompatible call pairs in GVN/NewGVN. However, I cannot provide negative tests for these changes.

Fixes https://github.com/llvm/llvm-project/issues/113997.

### CODE REVIEW

File: llvm/test/Transforms/GVN/pr113997.ll [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  %cond = select i1 %cmp1, i1 %cmp2, i1 false
+  br i1 %cond, label %if.then, label %if.else
+
+if.else:
+  %ctpop2 = tail call range(i32 0, 33) i32 @llvm.ctpop.i32(i32 %x)
+  %res = icmp eq i32 %ctpop2, 1
+  ret i1 %res
+
+if.then:
+  ret i1 false
```
Comment (andjo403) [inline llvm/test/Transforms/GVN/pr113997.ll:32]: what happen if there was a call to ctpop here also that can intersect with the entry block call but not with the if.else call?
Maybe not possible I do not know how this works.

Comment (dtcxzyw) [reply llvm/test/Transforms/GVN/pr113997.ll:32]: > what happen if there was a call to ctpop here also that can intersect with the entry block call but not with the if.else call?

In practice, it implies that the ctpop call in `%if.else` cannot intersect with the entry block call. Even if it happens, we do CSE first for calls in the entry and `%if.else`, and the second CSE will be rejected.

**Comment by goldsteinn** [comment]
This looks fine to me, although I think an easier way to implement this would be make `patchReplacementInstruction` return a bool to indicate `I` and `Repl` where successfully patched.

**Comment by dtcxzyw** [comment]
> an easier way to implement this would be make patchReplacementInstruction return a bool to indicate I and Repl where successfully patched.

I tried that before. But it may be too late in `GVNPass::performScalarPRE` :(

**Comment by pranavk** [comment]
LGTM. I verified that this patch fixes our test case that was broken earlier.

**Review by pranavk — APPROVED** [review]
LGTM but someone more knowledgable than me should also take a look.

**Comment by asmok-g** [comment]
We need this to unblock us in google. When can we expect a merge ?

**Comment by dtcxzyw** [comment]
> We need this to unblock us in google. When can we expect a merge ?

I will land this patch tomorrow if no more comments.

**Review by nikic — COMMENTED** [review]
The general implementation approach here looks incorrect to me. You can't really restrict replacements after the fact in GVN, you need to do so during value numbering.

Consider this minor variation of the test case:
```
define i1 @bucket2(i32 noundef %x) {
; CHECK-LABEL: define i1 @bucket2(
; CHECK-SAME: i32 noundef [[X:%.*]]) {
; CHECK-NEXT:    [[CMP1:%.*]] = icmp sgt i32 [[X]], 0
; CHECK-NEXT:    [[CTPOP1:%.*]] = tail call range(i32 1, 32) i32 @llvm.ctpop.i32(i32 zeroext [[X]])
; CHECK-NEXT:    [[CTPOP1INC:%.*]] = add i32 [[CTPOP1]], 1
; CHECK-NEXT:    [[CMP2:%.*]] = icmp samesign ult i32 [[CTPOP1INC]], 3
; CHECK-NEXT:    [[COND:%.*]] = select i1 [[CMP1]], i1 [[CMP2]], i1 false
; CHECK-NEXT:    br i1 [[COND]], label %[[IF_THEN:.*]], label %[[IF_ELSE:.*]]
; CHECK:       [[IF_ELSE]]:
; CHECK-NEXT:    [[CTPOP2:%.*]] = tail call range(i32 0, 33) i32 @llvm.ctpop.i32(i32 [[X]])
; CHECK-NEXT:    [[RES:%.*]] = icmp eq i32 [[CTPOP1INC]], 2
; CHECK-NEXT:    ret i1 [[RES]]
; CHECK:       [[IF_THEN]]:
; CHECK-NEXT:    ret i1 false
;
  %cmp1 = icmp sgt i32 %x, 0
  %ctpop1 = tail call range(i32 1, 32) i32 @llvm.ctpop.i32(i32 zeroext %x)
  %ctpop1inc = add i32 %ctpop1, 1
  %cmp2 = icmp samesign ult i32 %ctpop1inc, 3
  %cond = select i1 %cmp1, i1 %cmp2, i1 false
  br i1 %cond, label %if.then, label %if.else

if.else:
  %ctpop2 = tail call range(i32 0, 33) i32 @llvm.ctpop.i32(i32 %x)
  %ctpop2inc = add i32 %ctpop2, 1
  %res = icmp eq i32 %ctpop2inc, 2
  ret i1 %res

if.then:
  ret i1 false
}
```

This just adds a `zeroext` dummy attribute to make sure the two ctpop calls are not combined directly, and an extra `add 1` after them. Note that the adds *will* get CSEd, because the canBeReplacedBy check is specifically on the calls, not on any later instructions that are numbered based on the incorrect assumption that the calls are equivalent.

**Comment by dtcxzyw** [comment]
> You can't really restrict replacements after the fact in GVN, you need to do so during value numbering.

I've considered this approach before. But it is too heavy to add `Attributes` to `GVNPass::Expression`.

>  Note that the adds will get CSEd, because the canBeReplacedBy check is specifically on the calls, not on any later instructions that are numbered based on the incorrect assumption that the calls are equivalent.

You are right. I missed this case.

**Comment by nikic** [comment]
> > You can't really restrict replacements after the fact in GVN, you need to do so during value numbering.
> 
> I've considered this approach before. But it is too heavy to add `Attributes` to `GVNPass::Expression`.

I don't have any particularly good ideas on how to fix this properly. Given that this is a CSE transform, I think we could get away with only dropping the poison-generating attributes, and otherwise keeping the attributes of the first call, to avoid the problematic !canBeReplacedBy case.

**Pull request merged**